fabric-networking-api-v1/src/client/java/net/fabricmc/fabric/api/client/networking/v1/ClientPlayNetworking.java

```diff
@@ -255,4 +402,56 @@ public interface PlayChannelHandler {
 		 */
 		void receive(MinecraftClient client, ClientPlayNetworkHandler handler, PacketByteBuf buf, PacketSender responseSender);
 	}
+
+	/**
+	 * An internal packet handler that works as a proxy between old and new API.
+	 * @param <T> the type of the packet
+	 */
+	private interface PlayChannelHandlerProxy<T extends FabricPacket> extends PlayChannelHandler {
+		PlayPacketHandler<T> getOriginalHandler();
+	}
+
+	/**
+	 * A thread-safe packet handler utilizing {@link FabricPacket}.
+	 * @param <T> the type of the packet
+	 */
+	@FunctionalInterface
+	public interface PlayPacketHandler<T extends FabricPacket> {
+		/**
+		 * Handles the incoming packet. This is called on the render thread, and can safely
+		 * call client methods.
+		 *
+		 * <p>An example usage of this is to display an overlay message:
+		 * <pre>{@code
+		 * // See FabricPacket for creating the packet
+		 * ClientPlayNetworking.registerReceiver(OVERLAY_PACKET_TYPE, (player, packet, responseSender) -> {
+		 * 	MinecraftClient.getInstance().inGameHud.setOverlayMessage(packet.message(), true);
+		 * });
+		 * }</pre>
+		 *
+		 * <p>The network handler can be accessed via {@link ClientPlayerEntity#networkHandler}.
+		 *
+		 * @param packet the packet
+		 * @param player the player that received the packet
+		 * @param responseSender the packet sender
+		 * @see FabricPacket
+		 */
+		void receive(T packet, ClientPlayerEntity player, PacketSender responseSender);
+
+		/**
+		 * Processes an incoming packet. This is called on the network thread.
+		 * By default, this returns {@code true}, and {@link #receive} is called normally.
+		 * If this method returns {@code false}, {@link #receive} is not called.
+		 * This is for advanced users only; for normal uses, {@link #receive} should be
+		 * implemented instead.
+		 *
+		 * @param packet the packet
+		 * @param player the player that received the packet
+		 * @param responseSender the packet sender
+		 * @return {@code true} to proceed to {@link #receive}
+		 */
+		default boolean processOnNetworkThread(T packet, ClientPlayerEntity player, PacketSender responseSender) {
```

Comment:
Im a little unsure about the naming for this? Its fine unless someone can come up with something better.

Reply:
I agree, the name is the only blocker for me.

Reply:
Hmm, yeah, we need a better name but I can't really come up with this